src/services/keystore.ts

```diff
@@ -155,61 +172,162 @@ export class TrustedKeyStoreService {
         return this.storage.getItem(this.STORAGE_KEY_AUTO_FLAG) !== null;
     }
 
+    private encodePushToken(token: string | null, type: threema.PushTokenType | null): Uint8Array {
+        // Add prefix to push token string
+        let encodedString = null;
+        if (token !== null && type !== null) {
+            switch (type) {
+                case threema.PushTokenType.Fcm:
+                    encodedString = threema.PushTokenPrefix.Fcm + ':' + token;
+                    break;
+                case threema.PushTokenType.Apns:
+                    encodedString = threema.PushTokenPrefix.Apns + ':' + token;
+                    break;
+                case threema.PushTokenType.Hms:
+                    encodedString = threema.PushTokenPrefix.Hms + ':' + token;
+                    break;
+            }
+        }
+        return (encodedString == null)
+            ? new Uint8Array(0)
+            : (new TextEncoder()).encode(encodedString);
+    }
+
+    private decodePushToken(encodedBytes: Uint8Array): {token: string, type: threema.PushTokenType} {
+        const encodedString: string | null = encodedBytes.byteLength > 0
+            ? (new TextDecoder('utf-8', {fatal: true})).decode(encodedBytes)
+            : null;
+        let type: threema.PushTokenType = null;
+        let token: string = null;
+        if (encodedString !== null && encodedString[1] === ':') {
+            switch (encodedString[0]) {
+                case threema.PushTokenPrefix.Fcm:
+                    type = threema.PushTokenType.Fcm;
+                    break;
+                case threema.PushTokenPrefix.Apns:
+                    type = threema.PushTokenType.Apns;
+                    break;
+                case threema.PushTokenPrefix.Hms:
+                    type = threema.PushTokenType.Hms;
+                    break;
+                default:
+                    this.log.error('Invalid push token type:', encodedString[0]);
+                    return null;
+            }
+            token = encodedString.slice(2);
+        } else if (encodedString !== null) {
+            // Compat
+            token = encodedString;
+            type = threema.PushTokenType.Fcm;
+        }
+        return {token: token, type: type};
+    }
+
+    private async migrateDeprecatedStorageFormat(parts: [string, string], password: string): Promise<void> {
+        const nonce = hexToU8a(parts[0]);
+        const encrypted = hexToU8a(parts[1]);
+        const decrypted = nacl.secretbox.open(
+            encrypted,
+            nonce,
+            this.deprecatedPwToKey((new TextEncoder()).encode(password)),
+        );
+        if (!decrypted) {
+            throw new Error('Unable to decrypt deprecated storage');
+        }
+        const token = this.decodePushToken((decrypted as Uint8Array).slice(96));
```

Comment:
We should probably add a null check here to avoid an NPE (if decoding failed).

Reply:
I don't understand?

Reply:
`decodePushToken` may return null if decoding fails (for whatever reason). If that's the case, `token.token` and `token.type` below will result in a null pointer exception / type error.

Reply:
Okay. NPE confused me. ;) There's no exception, it would just be undefined (which is even worse).

Reply:
What I meant is that four lines below, `token.token` and `token.type` are accessed. And `(null).token` does result in an exception 🙂